tokio/src/runtime/task/join.rs

```diff
@@ -87,13 +104,171 @@ unsafe impl<T: Send> Sync for JoinHandle<T> {}
 impl<T> JoinHandle<T> {
     pub(super) fn new(raw: RawTask) -> JoinHandle<T> {
         JoinHandle {
-            raw: Some(raw),
+            raw: RawJoinHandle { raw: Some(raw) },
             _p: PhantomData,
         }
     }
+
+    /// Cancel the task associated with the handle.
+    ///
+    /// Awaiting a cancelled task might complete as usual if the task was
+    /// already completed at the time it was cancelled, but most likely it
+    /// will complete with a `Err(JoinError::Cancelled)`.
+    ///
+    /// ```rust
+    /// use tokio::time;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///    let mut handles = Vec::new();
+    ///
+    ///    handles.push(tokio::spawn(async {
+    ///       time::delay_for(time::Duration::from_secs(10)).await;
+    ///       true
+    ///    }));
+    ///
+    ///    handles.push(tokio::spawn(async {
+    ///       time::delay_for(time::Duration::from_secs(10)).await;
+    ///       false
+    ///    }));
+    ///
+    ///    for handle in &handles {
+    ///        handle.cancel();
+    ///    }
+    ///
+    ///    for handle in handles {
+    ///        assert!(handle.await.unwrap_err().is_cancelled());
+    ///    }
+    /// }
+    /// ```
+    pub fn cancel(&self) {
+        if let Some(raw) = self.raw.raw {
+            raw.shutdown();
+        }
+    }
+
+    /// Convert the join handle into a raw join handle.
+    ///
+    /// Raw join handles erase the type information associated with a
+    /// [`JoinHandle`], allowing them to easily be stored in containers for
+    /// future cancellation.
+    ///
+    /// [`JoinHandle`]: JoinHandle
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use tokio::time;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///    let mut handles = Vec::new();
+    ///
+    ///    let handle = tokio::spawn(async {
+    ///       time::delay_for(time::Duration::from_secs(10)).await;
+    ///       true
+    ///    });
+    ///
+    ///    handles.push(handle.into_raw_handle());
+    ///
+    ///    let handle = tokio::spawn(async {
+    ///       time::delay_for(time::Duration::from_secs(10)).await;
+    ///       1u32
+    ///    });
+    ///
+    ///    handles.push(handle.into_raw_handle());
+    ///
+    ///    for handle in &handles {
+    ///        handle.cancel();
+    ///    }
+    ///
+    ///    for handle in handles {
+    ///        assert!(handle.await.unwrap_err().is_cancelled());
+    ///    }
+    /// }
+    /// ```
+    pub fn into_raw_handle(self) -> RawJoinHandle {
```

Comment:
I'm not necessarily against this, but could we decouple exposing `RawJoinHandle` as a public API from this PR? I would like to explore alternate solutions like providing a custom collection type.

Reply:
Rebased and renamed to `abort` for now.

How do you want me to hide this from the public API?

Completely remove it, make private with `#[allow(unused)]`, or `#[doc(hidden)]`?

Reply:
Completely remove it in this PR. I would like to get abort landed. I don’t think a raw handle is the path forward. Lets discuss the use cases that motivated you that direction so we can figure out an option.

Reply:
Removed, thanks!